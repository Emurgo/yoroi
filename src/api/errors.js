// @flow
import ExtendableError from 'es6-error'

// thrown when the request did go through but
// backend returned an unexpected result
export class ApiError extends ExtendableError {
  constructor(request: any) {
    super('ApiError')
    this.request = request
  }
}

// thrown when api failed to connect to the server
export class NetworkError extends ExtendableError {
  constructor() {
    super('NetworkError')
  }
}

// thrown by the backend after a rollback
export class ApiHistoryError extends ApiError {
  constructor(status: string) {
    super(`ApiHistoryError::${status}`)
  }
}
