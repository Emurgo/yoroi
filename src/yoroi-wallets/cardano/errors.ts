import ExtendableError from 'es6-error'

import {errorMessages} from '../../i18n/global-messages'
import LocalizableError from '../../i18n/LocalizableError'

// thrown when the request did go through but
// backend returned an unexpected result
export class ApiError extends LocalizableError {
  constructor(response: string | null) {
    super({
      id: errorMessages.apiError.message.id,
      defaultMessage: errorMessages.apiError.message.defaultMessage || '',
      values: {
        response,
      },
    })
  }
}

// thrown when api failed to connect to the server
export class NetworkError extends LocalizableError {
  constructor() {
    super({
      id: errorMessages.networkError.message.id,
      defaultMessage: errorMessages.networkError.message.defaultMessage || '',
    })
  }
}

// thrown by the backend after a rollback
export class ApiHistoryError extends ApiError {
  public static readonly errors = {
    REFERENCE_TX_NOT_FOUND: 'REFERENCE_TX_NOT_FOUND',
    REFERENCE_BLOCK_MISMATCH: 'REFERENCE_BLOCK_MISMATCH',
    REFERENCE_BEST_BLOCK_MISMATCH: 'REFERENCE_BEST_BLOCK_MISMATCH',
  }
}

export class CardanoError extends ExtendableError {}
export class WrongPassword extends ExtendableError {
  constructor() {
    super('WrongPassword')
  }
}
export class InvalidState extends ExtendableError {}
