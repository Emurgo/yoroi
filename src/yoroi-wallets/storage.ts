import AsyncStorage, {AsyncStorageStatic} from '@react-native-async-storage/async-storage'

export type Storage = {
  getItem: AsyncStorageStatic['getItem']
  multiGet: AsyncStorageStatic['multiGet']
  setItem: AsyncStorageStatic['setItem']
  multiSet: AsyncStorageStatic['multiSet']
  removeItem: AsyncStorageStatic['removeItem']
  multiRemove: AsyncStorageStatic['multiRemove']
}

export const makeStorageWithPrefix = (prefix: string): Storage => {
  const withPrefix = (key: string) => `${prefix}/${key}`

  return {
    getItem: (key: string) => {
      return AsyncStorage.getItem(withPrefix(key))
    },
    multiGet: (keys: Array<string>) => {
      return AsyncStorage.multiGet(keys.map((key) => withPrefix(key)))
    },
    setItem: (key: string, value: string) => {
      return AsyncStorage.setItem(withPrefix(key), value)
    },
    multiSet: (items: Array<[string, string]>) => {
      return AsyncStorage.multiSet(items.map(([key, value]) => [withPrefix(key), value]))
    },
    removeItem: (key: string) => {
      return AsyncStorage.removeItem(withPrefix(key))
    },
    multiRemove: (keys: Array<string>) => {
      return AsyncStorage.multiRemove(keys.map((key) => withPrefix(key)))
    },
  }
}
