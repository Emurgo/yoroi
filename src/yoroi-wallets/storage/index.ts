export * from './EncryptedStorage'
export * from './Keychain'
export * from './KeychainStorage'
export * from './storage'
export * from './StorageProvider'
