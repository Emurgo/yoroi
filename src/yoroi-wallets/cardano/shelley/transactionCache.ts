/* eslint-disable @typescript-eslint/no-explicit-any */
import {fromPairs, mapValues, max, sum, uniq} from 'lodash'
import {defaultMemoize} from 'reselect'

import * as api from '../../../legacy/api'
import assert from '../../../legacy/assert'
import {CONFIG} from '../../../legacy/config'
import {ApiHistoryError} from '../../../legacy/errors'
import {ObjectValues} from '../../../legacy/flow'
import type {Transaction} from '../../../legacy/HistoryTransaction'
import {TRANSACTION_STATUS} from '../../../legacy/HistoryTransaction'
import {Logger} from '../../../legacy/logging'
import {getCardanoNetworkConfigById} from '../../../legacy/networks'
import {limitConcurrency} from '../../../legacy/promise'
import type {NetworkId, YoroiProvider} from '../../../legacy/types'
import type {RemoteCertificateMeta, TxHistoryRequest} from '../../../legacy/types'
import {CERTIFICATE_KIND} from '../../../legacy/types'

type SyncMetadata = {
  bestBlockNum: number
  bestBlockHash: string | null | undefined
  bestTxHash: string | null | undefined
}
type TransactionCacheState = {
  transactions: Record<string, Transaction>
  perAddressSyncMetadata: Record<string, SyncMetadata>
  bestBlockNum: number | null | undefined // global best block, not per address
}
export type TimeForTx = {
  blockHash: string
  blockNum: number
  txHash: string
  txOrdinal: number
}

const getLatestTransaction: (arg0: Array<Transaction>) => void | TimeForTx = (txs) => {
  const blockInfo: Array<TimeForTx> = []

  for (const tx of txs) {
    if (tx.blockHash != null && tx.txOrdinal != null && tx.blockNum != null) {
      blockInfo.push({
        blockHash: tx.blockHash,
        txHash: tx.id,
        txOrdinal: tx.txOrdinal,
        blockNum: tx.blockNum,
      })
    }
  }

  if (blockInfo.length === 0) {
    return undefined
  }

  let best = blockInfo[0]

  for (let i = 1; i < blockInfo.length; i++) {
    if (blockInfo[i].blockNum > best.blockNum) {
      best = blockInfo[i]
      continue
    }

    if (blockInfo[i].blockNum === best.blockNum) {
      if (blockInfo[i].txOrdinal > best.txOrdinal) {
        best = blockInfo[i]
        continue
      }
    }
  }

  return best
}

const perAddressTxsSelector = (state: TransactionCacheState) => {
  const transactions = state.transactions
  const addressToTxs = {}

  const addTxTo = (txId, addr) => {
    const current = addressToTxs[addr] || []
    const cleared = current.filter((_txId) => txId !== _txId)
    addressToTxs[addr] = [...cleared, txId]
  }

  ObjectValues(transactions).forEach((tx) => {
    tx.inputs.forEach(({address}) => {
      addTxTo(tx.id, address)
    })
    tx.outputs.forEach(({address}) => {
      addTxTo(tx.id, address)
    })
  })
  return addressToTxs
}

export type TimestampedCertMeta = {
  submittedAt: string
  epoch: number
  certificates: Array<RemoteCertificateMeta>
}
type PerAddressCertificatesDict = Record<string, Record<string, TimestampedCertMeta>>

const perAddressCertificatesSelector = (state: TransactionCacheState): PerAddressCertificatesDict => {
  const transactions = state.transactions
  const addressToPerTxCerts = {}

  const addTxTo = (
    txId: string,
    certificates: Array<RemoteCertificateMeta>,
    submittedAt: string | null | undefined,
    epoch: number | null | undefined,
    addr: string,
  ) => {
    const current: Record<string, TimestampedCertMeta> = addressToPerTxCerts[addr] || {}

    if (current[txId] == null && submittedAt != null && epoch != null) {
      current[txId] = {
        submittedAt,
        epoch,
        certificates,
      }
      addressToPerTxCerts[addr] = current
    }
  }

  ObjectValues(transactions).forEach((tx) => {
    tx.certificates.forEach((cert) => {
      if (
        cert.kind === CERTIFICATE_KIND.STAKE_REGISTRATION ||
        cert.kind === CERTIFICATE_KIND.STAKE_DEREGISTRATION ||
        cert.kind === CERTIFICATE_KIND.STAKE_DELEGATION
      ) {
        const {rewardAddress} = cert as any
        addTxTo(tx.id, tx.certificates, tx.submittedAt, tx.epoch, rewardAddress)
      }
    })
  })
  return addressToPerTxCerts
}

const confirmationCountsSelector = (state: TransactionCacheState) => {
  const {perAddressSyncMetadata, transactions} = state
  return mapValues(transactions, (tx: Transaction) => {
    if (tx.status !== TRANSACTION_STATUS.SUCCESSFUL) {
      // TODO(ppershing): do failed transactions have assurance?
      return null
    }

    const getBlockNum = ({address}) =>
      perAddressSyncMetadata[address] ? perAddressSyncMetadata[address].bestBlockNum : 0

    const bestBlockNum: any = max([
      state.bestBlockNum || 0,
      ...tx.inputs.map(getBlockNum),
      ...tx.outputs.map(getBlockNum),
    ])
    assert.assert(tx.blockNum, 'Successfull tx should have blockNum')

    return bestBlockNum - (tx as any).blockNum
  })
}

export type TransactionCacheJSON = TransactionCacheState
export class TransactionCache {
  _state: TransactionCacheState = {
    perAddressSyncMetadata: {},
    transactions: {},
    bestBlockNum: 0,
  }

  _subscriptions: Array<() => any> = []
  _onTxHistoryUpdateSubscriptions: Array<() => any> = []
  _perAddressTxsSelector = defaultMemoize(perAddressTxsSelector)
  _perAddressCertificates = defaultMemoize(perAddressCertificatesSelector)
  _confirmationCountsSelector = defaultMemoize(confirmationCountsSelector)

  subscribe(handler: () => any) {
    this._subscriptions.push(handler)
  }

  notifyOnTxHistoryUpdate = () => {
    this._onTxHistoryUpdateSubscriptions.forEach((handler) => handler())
  }

  subscribeOnTxHistoryUpdate(handler: () => any) {
    this._onTxHistoryUpdateSubscriptions.push(handler)
  }

  updateState(update: TransactionCacheState) {
    Logger.debug('TransactionHistory update state')
    // Logger.debug('Update', update)
    this._state = {...this._state, ...update}

    this._subscriptions.forEach((handler) => handler())
  }

  resetState() {
    this.updateState({
      perAddressSyncMetadata: {},
      transactions: {},
      bestBlockNum: 0,
    })
  }

  get transactions() {
    return this._state.transactions
  }

  get perAddressTxs() {
    return this._perAddressTxsSelector(this._state)
  }

  get perRewardAddressCertificates() {
    return this._perAddressCertificates(this._state)
  }

  get confirmationCounts() {
    return this._confirmationCountsSelector(this._state)
  }

  _getBlockMetadata(addrs: Array<string>) {
    assert.assert(addrs.length, 'getBlockMetadata: addrs not empty')
    const metadata = addrs.map((addr) => this._state.perAddressSyncMetadata[addr])
    const first = metadata[0]

    if (!first) {
      // New addresses
      assert.assert(
        metadata.every((x) => !x),
        'getBlockMetadata: undefined vs defined',
      )
      return {
        bestBlockNum: 0,
        bestBlockHash: null,
        bestTxHash: null,
      }
    } else {
      // Old addresses
      assert.assert(
        metadata.every((x) => x.bestBlockNum === first.bestBlockNum),
        'getBlockMetadata: bestBlockNum metadata same',
      )
      assert.assert(
        metadata.every((x) => x.bestBlockHash === first.bestBlockHash),
        'getBlockMetadata: bestBlockHash metadata same',
      )
      return first
    }
  }

  _buildTxHistoryRequest(
    addresses: Array<string>,
    metadata: SyncMetadata,
    currentBestBlockHash: string | null | undefined,
  ): TxHistoryRequest {
    assert.assert(currentBestBlockHash != null, 'buildTxHistoryRequest: bestBlock not null')

    /* :: if (currentBestBlockHash == null) throw 'assert' */
    const request = {
      addresses,
      untilBlock: currentBestBlockHash,
    }

    if (metadata.bestBlockHash != null && metadata.bestTxHash != null) {
      return {
        ...request,
        after: {
          block: metadata.bestBlockHash,
          tx: metadata.bestTxHash,
        },
      } as any
    }

    return request as any
  }

  _isUpdatedTransaction(tx: Transaction): boolean {
    const id = tx.id

    // We have this transaction and it did not change
    if (this._state.transactions[id] && this._state.transactions[id].lastUpdatedAt === tx.lastUpdatedAt) {
      return false
    }

    if (this._state.transactions[id]) {
      // Do things that matter if the transaction changed!
      Logger.info('Tx changed', tx)
    }

    return true
  }

  // Returns number of updated transactions
  _checkUpdatedTransactions(transactions: Array<Transaction>): number {
    // Currently we do not support two updates inside a same batch
    // (and backend shouldn't support it either)
    assert.assert(
      transactions.length === uniq(transactions.map((tx) => tx.id)).length,
      'Got the same transaction twice in one batch',
      transactions,
    )
    const updated = transactions.map((tx) => this._isUpdatedTransaction(tx))

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return (sum as any)(updated, (x) => (x ? 1 : 0))
  }

  async doSyncStep(
    blocks: Array<Array<string>>,
    networkId: NetworkId,
    provider: YoroiProvider | null | undefined,
  ): Promise<boolean> {
    let count = 0
    let wasPaginated = false
    const errors: Array<Error> = []
    const currentBestBlock = await api.getBestBlock(getCardanoNetworkConfigById(networkId, provider).BACKEND)
    const tasks = blocks.map((addrs) => {
      const metadata = this._getBlockMetadata(addrs)

      const historyRequest = this._buildTxHistoryRequest(addrs, metadata, currentBestBlock.hash)

      return () =>
        api
          .fetchNewTxHistory(historyRequest, getCardanoNetworkConfigById(networkId, provider).BACKEND)
          .then((response) => [addrs, response] as const)
    })
    const limit = limitConcurrency(CONFIG.MAX_CONCURRENT_REQUESTS)
    const promises = tasks.map((t) => limit(t))

    // Note(ppershing): This serializes the response order
    // but still allows for concurrent requests
    for (const promise of promises) {
      try {
        const [addrs, response] = await promise
        wasPaginated = wasPaginated || !response.isLast

        const metadata = this._getBlockMetadata(addrs)

        const bestTx = getLatestTransaction(response.transactions)
        // Note: we can update best block number only if we are processing
        // the last page of the history request, see design doc for details
        const newBestBlockNum =
          response.isLast && response.transactions.length ? currentBestBlock.height : metadata.bestBlockNum
        const newMetadata = {
          bestBlockNum: newBestBlockNum,
          bestBlockHash: bestTx?.blockHash,
          bestTxHash: bestTx?.txHash,
        }
        const transactionsUpdate = fromPairs(response.transactions.map((tx) => [tx.id, tx]))
        const metadataUpdate = fromPairs(addrs.map((addr) => [addr, newMetadata]))
        count += this._checkUpdatedTransactions(response.transactions)

        if (count > 0) {
          Logger.info('TransactionCache: notifying subscribers on txs update')
          this.notifyOnTxHistoryUpdate()
        }

        this.updateState({
          transactions: {
            ...this._state.transactions,
            ...transactionsUpdate,
          } as any,
          perAddressSyncMetadata: {
            ...this._state.perAddressSyncMetadata,
            ...metadataUpdate,
          } as any,
          bestBlockNum: currentBestBlock.height,
        })
      } catch (e) {
        if (e instanceof ApiHistoryError) {
          // flush cache to resync
          // (consider only removing cache for current address block)
          this.resetState()
          throw e
        }

        errors.push(e as Error)
      }
    }

    if (errors.length) {
      Logger.error('Sync errors', errors)
      throw errors
    }

    return wasPaginated || count > 0
  }

  toJSON(): TransactionCacheJSON {
    return this._state
  }

  static fromJSON(data: TransactionCacheJSON) {
    const cache = new TransactionCache()
    // if cache is deprecated it means it was obtained when the old history
    // endpoint was still being used (in versions <= 2.2.1)
    // in this case, we do not load the data and start from a fresh object.
    const isDeprecatedCache = ObjectValues(data.perAddressSyncMetadata).some(
      (metadata: any) => metadata.lastUpdated != null,
    )
    // similarly, the haskell shelley endpoint introduces withdrawals & certs.
    // versions < 3.0.1 need resync
    const txs = ObjectValues(data.transactions)
    const lacksShelleyTxData = txs ? txs.some((tx) => tx.withdrawals == null || tx.certificates == null) : false

    if (!isDeprecatedCache && !lacksShelleyTxData) {
      cache.updateState(data)
    } else {
      Logger.info('TransactionCache: deprecated tx. Restoring from empty object')
    }

    return cache
  }
}
