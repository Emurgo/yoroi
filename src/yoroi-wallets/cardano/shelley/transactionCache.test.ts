import {fromPairs} from 'lodash'
import DeviceInfo from 'react-native-device-info'

import {ApiHistoryError} from '../../../legacy/errors'
import {storage as rootStorage, YoroiStorage} from '../../storage'
import type {Transaction} from '../../types/other'
import {
  mockedAddressesByChunks,
  mockedBackendConfig,
  mockedEmptyHistoryResponse,
  mockedEmptyLocalTransactions,
  mockedHistoryResponse,
  mockedLocalTransactions,
  mockedTipStatusResponse,
  mockTx,
} from '../mocks'
import {makeTxCacheStorage, syncTxs, toCachedTx, TransactionCache} from './transactionCache'

describe('transactionCache', () => {
  describe('create', () => {
    it('loads from storage', async () => {
      DeviceInfo.getVersion = () => '9.9.9'
      const txCache = await TransactionCache.create(mockStorage)

      expect(txCache.transactions).toMatchSnapshot()
      expect(txCache.perAddressTxs).toMatchSnapshot()
      expect(txCache.perRewardAddressCertificates).toMatchSnapshot()
      expect(txCache.confirmationCounts).toMatchSnapshot()

      txCache.resetState()

      expect(txCache.transactions).toMatchSnapshot()
      expect(txCache.perAddressTxs).toMatchSnapshot()
      expect(txCache.perRewardAddressCertificates).toMatchSnapshot()
      expect(txCache.confirmationCounts).toMatchSnapshot()
    })

    it('starts fresh if txids is invalid format', async () => {
      DeviceInfo.getVersion = () => '9.9.9'

      const txCache = await TransactionCache.create({
        ...mockStorage,
        getItem: async (path: string) => {
          if (path !== 'txids') {
            throw new Error('invalid path')
          }
          return undefined // invalid txids
        },
      })

      expect(txCache.transactions).toMatchSnapshot()
      expect(txCache.perAddressTxs).toMatchSnapshot()
      expect(txCache.perRewardAddressCertificates).toMatchSnapshot()
      expect(txCache.confirmationCounts).toMatchSnapshot()
    })

    it('drops transaction if invalid format', async () => {
      DeviceInfo.getVersion = () => '9.9.9'
      const txCache = await TransactionCache.create({
        ...mockStorage,
        multiGet: async (paths: Array<string>) => {
          if (paths.length === 1 && paths[0] === mockTx.id) {
            return [[mockTx.id, undefined] as [string, undefined]]
          }
          throw new Error('invalid path')
        },
      })

      expect(txCache.transactions).toMatchSnapshot()
      expect(txCache.perAddressTxs).toMatchSnapshot()
      expect(txCache.perRewardAddressCertificates).toMatchSnapshot()
      expect(txCache.confirmationCounts).toMatchSnapshot()
    })

    it('starts fresh if upgrading from old version', async () => {
      DeviceInfo.getVersion = () => '0.0.1'
      const txCache = await TransactionCache.create(mockStorage)

      expect(txCache.transactions).toMatchSnapshot()
      expect(txCache.perAddressTxs).toMatchSnapshot()
      expect(txCache.perRewardAddressCertificates).toMatchSnapshot()
      expect(txCache.confirmationCounts).toMatchSnapshot()
    })
  })
})

describe('transaction storage', () => {
  it('works', async () => {
    const storage = rootStorage.join('txs/')
    const {loadTxs, saveTxs, clear} = makeTxCacheStorage(storage)

    // initial
    await loadTxs().then((txs) => {
      return expect(txs).toEqual({})
    })

    await saveTxs({[mockTx.id]: mockTx})
    await loadTxs().then((txs) => {
      return expect(txs).toEqual({[mockTx.id]: mockTx})
    })

    await clear()
    await loadTxs().then((txs) => {
      return expect(txs).toEqual({})
    })
  })
})

const mockStorage: YoroiStorage = {
  getItem: async (path: string) => {
    if (path === 'txids') return [mockTx.id]
    throw new Error('invalid path')
  },
  multiGet: async (txids: Array<string>) => {
    if (txids[0] !== mockTx.id) throw new Error('invalid path')

    return [
      [txids[0], mockTx] as [string, Transaction], //
    ]
  },
  join: () => mockStorage,
  setItem: jest.fn(),
  multiSet: jest.fn(),
  removeItem: jest.fn(),
  multiRemove: jest.fn(),
  getAllKeys: jest.fn(),
  clear: jest.fn(),
}

describe('syncTxs (undefined means no updates)', () => {
  it('should return undefined if tipStatus bestBlock.hash is empty', async () => {
    const params = {
      addressesByChunks: [],
      backendConfig: mockedBackendConfig,
      transactions: mockedEmptyLocalTransactions,
      api: {
        getTipStatus: jest.fn().mockResolvedValue({
          safeBlock: mockedTipStatusResponse.safeBlock,
          bestBlock: {...mockedTipStatusResponse.bestBlock, hash: ''},
        }),
        fetchNewTxHistory: jest.fn(),
      },
    }

    const result = await syncTxs(params)

    expect(result).toBeUndefined()
    expect(params.api.getTipStatus).toBeCalledTimes(1)
  })

  it('should return undefined if there is no new transactions', async () => {
    const params = {
      addressesByChunks: mockedAddressesByChunks,
      backendConfig: mockedBackendConfig,
      transactions: mockedLocalTransactions,
      api: {
        getTipStatus: jest.fn().mockResolvedValue(mockedTipStatusResponse),
        fetchNewTxHistory: jest
          .fn()
          .mockResolvedValueOnce(mockedEmptyHistoryResponse)
          .mockResolvedValueOnce(mockedEmptyHistoryResponse)
          .mockResolvedValueOnce(mockedEmptyHistoryResponse),
      },
    }

    const result = await syncTxs(params)

    expect(result).toBeUndefined()
    expect(params.api.fetchNewTxHistory).toBeCalledTimes(3)
  })

  it('should return current txs plus new txs if there are new transactions', async () => {
    const params = {
      addressesByChunks: mockedAddressesByChunks,
      backendConfig: mockedBackendConfig,
      transactions: mockedLocalTransactions,
      api: {
        getTipStatus: jest.fn().mockResolvedValue(mockedTipStatusResponse),
        fetchNewTxHistory: jest
          .fn()
          .mockResolvedValueOnce(mockedHistoryResponse)
          .mockResolvedValueOnce(mockedEmptyHistoryResponse)
          .mockResolvedValueOnce(mockedEmptyHistoryResponse),
      },
    }
    const response = {
      ...mockedLocalTransactions,
      ...fromPairs(mockedHistoryResponse.transactions.map((t) => [t.hash, toCachedTx(t)])),
    }

    const result = await syncTxs(params)

    expect(result).toEqual(response)
    expect(params.api.fetchNewTxHistory).toBeCalledTimes(3)
  })

  it('should return current txs plus new txs if there are new transactions and continue to request while is not the last', async () => {
    const params = {
      addressesByChunks: [mockedAddressesByChunks[0]],
      backendConfig: mockedBackendConfig,
      transactions: {},
      api: {
        getTipStatus: jest.fn().mockResolvedValue(mockedTipStatusResponse),
        fetchNewTxHistory: jest
          .fn()
          .mockResolvedValueOnce({...mockedHistoryResponse, isLast: false})
          .mockResolvedValueOnce({...mockedHistoryResponse, isLast: false})
          .mockResolvedValueOnce({...mockedHistoryResponse, isLast: false})
          .mockResolvedValueOnce({...mockedHistoryResponse, isLast: false})
          .mockResolvedValueOnce(mockedEmptyHistoryResponse),
      },
    }
    const response = fromPairs(mockedHistoryResponse.transactions.map((t) => [t.hash, toCachedTx(t)]))

    const result = await syncTxs(params)

    expect(result).toEqual(response)
    expect(params.api.fetchNewTxHistory).toBeCalledTimes(5)
  })

  it.each([ApiHistoryError.errors.REFERENCE_BLOCK_MISMATCH, ApiHistoryError.errors.REFERENCE_TX_NOT_FOUND])(
    `should return current txs minus txs after last_tx.height if receives %p`,
    async (error) => {
      const params = {
        addressesByChunks: mockedAddressesByChunks,
        backendConfig: mockedBackendConfig,
        transactions: {
          ...mockedLocalTransactions,
          ...fromPairs(mockedHistoryResponse.transactions.map((t) => [t.hash, toCachedTx(t)])),
        },
        api: {
          getTipStatus: jest.fn().mockResolvedValue(mockedTipStatusResponse),
          fetchNewTxHistory: jest
            .fn()
            .mockRejectedValueOnce(new ApiHistoryError(error))
            .mockResolvedValueOnce(mockedEmptyHistoryResponse)
            .mockResolvedValueOnce(mockedEmptyHistoryResponse),
        },
      }

      const result = await syncTxs(params)

      expect(result).toEqual(mockedLocalTransactions)
    },
  )

  it(`should return undefined if receives ${ApiHistoryError.errors.REFERENCE_BEST_BLOCK_MISMATCH}`, async () => {
    const params = {
      addressesByChunks: mockedAddressesByChunks,
      backendConfig: mockedBackendConfig,
      transactions: mockedLocalTransactions,
      api: {
        getTipStatus: jest.fn().mockResolvedValue(mockedTipStatusResponse),
        fetchNewTxHistory: jest
          .fn()
          .mockRejectedValueOnce(new ApiHistoryError(ApiHistoryError.errors.REFERENCE_BEST_BLOCK_MISMATCH))
          .mockResolvedValueOnce(mockedEmptyHistoryResponse)
          .mockResolvedValueOnce(mockedEmptyHistoryResponse),
      },
    }

    const result = await syncTxs(params)

    expect(result).toBeUndefined()
  })

  it(`should return undefined if receives any other error`, async () => {
    const params = {
      addressesByChunks: mockedAddressesByChunks,
      backendConfig: mockedBackendConfig,
      transactions: mockedLocalTransactions,
      api: {
        getTipStatus: jest.fn().mockResolvedValue(mockedTipStatusResponse),
        fetchNewTxHistory: jest
          .fn()
          .mockRejectedValueOnce(new Error('error'))
          .mockResolvedValueOnce(mockedEmptyHistoryResponse)
          .mockResolvedValueOnce(mockedEmptyHistoryResponse),
      },
    }

    const result = await syncTxs(params)

    expect(result).toBeUndefined()
  })
})
