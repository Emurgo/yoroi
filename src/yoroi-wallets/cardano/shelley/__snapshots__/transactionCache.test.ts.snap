// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transactionCache create drops transaction if invalid format 1`] = `Object {}`;

exports[`transactionCache create drops transaction if invalid format 2`] = `Object {}`;

exports[`transactionCache create drops transaction if invalid format 3`] = `Object {}`;

exports[`transactionCache create drops transaction if invalid format 4`] = `Object {}`;

exports[`transactionCache create loads from storage 1`] = `
Object {
  "0a8962dde362eef1f840defe6f916fdf9701ad53c7cb5dd4a74ab85df8e9bffc": Object {
    "blockHash": "fb418acaa29c66e799a16b594f7beedfe2ef53413e9863b61a418f2df1ff1442",
    "blockNum": 2909238,
    "certificates": Array [
      Object {
        "kind": "StakeDeregistration",
        "rewardAddress": "e0acab7e493ece4c1e6ae627ef9f5f7c9b1063e599e4aa91f87f0d58ae",
      },
    ],
    "collateralInputs": Array [],
    "epoch": 156,
    "fee": "179537",
    "id": "0a8962dde362eef1f840defe6f916fdf9701ad53c7cb5dd4a74ab85df8e9bffc",
    "inputs": Array [
      Object {
        "address": "addr_test1qrrdv3uxj8shu27ea9djvnn3rl4w3lvh3cyck6yc36mvf6ctlqxj9g0azvpycncr9u600p6t556qhc3psk06uzzw6saq4kvdpq",
        "amount": "967141533",
        "assets": Array [
          Object {
            "amount": "1",
            "assetId": "57e37bc9a9c0a099a6636c3deb93b82e7edec8a9a40883017bae2674.717171717171",
            "name": "717171717171",
            "policyId": "57e37bc9a9c0a099a6636c3deb93b82e7edec8a9a40883017bae2674",
          },
          Object {
            "amount": "1",
            "assetId": "fc53320cfda5add9cde1e7094c73596eacc26dbe79834b67c14b5dad.656565656565",
            "name": "656565656565",
            "policyId": "fc53320cfda5add9cde1e7094c73596eacc26dbe79834b67c14b5dad",
          },
        ],
      },
      Object {
        "address": "addr_test1qqgxd3r59psq0dg33t7asmvjmtu55tvvcmeq5kmhj0tmqjctlqxj9g0azvpycncr9u600p6t556qhc3psk06uzzw6saqk4x7z6",
        "amount": "2000000",
        "assets": Array [
          Object {
            "amount": "1",
            "assetId": "0b71c073fcf017eeff0664070c790a2bcc47077566904be471c46c13.727272727272",
            "name": "727272727272",
            "policyId": "0b71c073fcf017eeff0664070c790a2bcc47077566904be471c46c13",
          },
        ],
      },
    ],
    "lastUpdatedAt": "2021-09-13T18:42:10.000Z",
    "memo": null,
    "outputs": Array [
      Object {
        "address": "addr_test1qrxlnftwl73taxvcapnhgctae895l582a6r7k7jjeuwvzp0rvvww4m29k4km54utxag3mlhdsr73m62rsae6ad3hj6kqcexkh8",
        "amount": "7305977",
        "assets": Array [],
      },
      Object {
        "address": "addr_test1qrqzse20fh7mmt5k9xf4sug3a2lh5fa7x9nr98avp0ac78stlqxj9g0azvpycncr9u600p6t556qhc3psk06uzzw6saq6xr7ra",
        "amount": "961656019",
        "assets": Array [
          Object {
            "amount": "1",
            "assetId": "0b71c073fcf017eeff0664070c790a2bcc47077566904be471c46c13.727272727272",
            "name": "727272727272",
            "policyId": "0b71c073fcf017eeff0664070c790a2bcc47077566904be471c46c13",
          },
          Object {
            "amount": "1",
            "assetId": "57e37bc9a9c0a099a6636c3deb93b82e7edec8a9a40883017bae2674.717171717171",
            "name": "717171717171",
            "policyId": "57e37bc9a9c0a099a6636c3deb93b82e7edec8a9a40883017bae2674",
          },
          Object {
            "amount": "1",
            "assetId": "fc53320cfda5add9cde1e7094c73596eacc26dbe79834b67c14b5dad.656565656565",
            "name": "656565656565",
            "policyId": "fc53320cfda5add9cde1e7094c73596eacc26dbe79834b67c14b5dad",
          },
        ],
      },
    ],
    "scriptSize": 0,
    "slot": 166914,
    "status": "Successful",
    "submittedAt": "2021-09-13T18:42:10.000Z",
    "txOrdinal": 0,
    "type": "shelley",
    "validContract": true,
    "withdrawals": Array [],
  },
}
`;

exports[`transactionCache create loads from storage 2`] = `
Object {
  "addr_test1qqgxd3r59psq0dg33t7asmvjmtu55tvvcmeq5kmhj0tmqjctlqxj9g0azvpycncr9u600p6t556qhc3psk06uzzw6saqk4x7z6": Array [
    "0a8962dde362eef1f840defe6f916fdf9701ad53c7cb5dd4a74ab85df8e9bffc",
  ],
  "addr_test1qrqzse20fh7mmt5k9xf4sug3a2lh5fa7x9nr98avp0ac78stlqxj9g0azvpycncr9u600p6t556qhc3psk06uzzw6saq6xr7ra": Array [
    "0a8962dde362eef1f840defe6f916fdf9701ad53c7cb5dd4a74ab85df8e9bffc",
  ],
  "addr_test1qrrdv3uxj8shu27ea9djvnn3rl4w3lvh3cyck6yc36mvf6ctlqxj9g0azvpycncr9u600p6t556qhc3psk06uzzw6saq4kvdpq": Array [
    "0a8962dde362eef1f840defe6f916fdf9701ad53c7cb5dd4a74ab85df8e9bffc",
  ],
  "addr_test1qrxlnftwl73taxvcapnhgctae895l582a6r7k7jjeuwvzp0rvvww4m29k4km54utxag3mlhdsr73m62rsae6ad3hj6kqcexkh8": Array [
    "0a8962dde362eef1f840defe6f916fdf9701ad53c7cb5dd4a74ab85df8e9bffc",
  ],
}
`;

exports[`transactionCache create loads from storage 3`] = `
Object {
  "e0acab7e493ece4c1e6ae627ef9f5f7c9b1063e599e4aa91f87f0d58ae": Object {
    "0a8962dde362eef1f840defe6f916fdf9701ad53c7cb5dd4a74ab85df8e9bffc": Object {
      "certificates": Array [
        Object {
          "kind": "StakeDeregistration",
          "rewardAddress": "e0acab7e493ece4c1e6ae627ef9f5f7c9b1063e599e4aa91f87f0d58ae",
        },
      ],
      "epoch": 156,
      "submittedAt": "2021-09-13T18:42:10.000Z",
    },
  },
}
`;

exports[`transactionCache create loads from storage 4`] = `
Object {
  "0a8962dde362eef1f840defe6f916fdf9701ad53c7cb5dd4a74ab85df8e9bffc": -2909238,
}
`;

exports[`transactionCache create loads from storage 5`] = `Object {}`;

exports[`transactionCache create loads from storage 6`] = `Object {}`;

exports[`transactionCache create loads from storage 7`] = `Object {}`;

exports[`transactionCache create loads from storage 8`] = `Object {}`;

exports[`transactionCache create starts fresh if txids is invalid format 1`] = `Object {}`;

exports[`transactionCache create starts fresh if txids is invalid format 2`] = `Object {}`;

exports[`transactionCache create starts fresh if txids is invalid format 3`] = `Object {}`;

exports[`transactionCache create starts fresh if txids is invalid format 4`] = `Object {}`;

exports[`transactionCache create starts fresh if upgrading from old version 1`] = `Object {}`;

exports[`transactionCache create starts fresh if upgrading from old version 2`] = `Object {}`;

exports[`transactionCache create starts fresh if upgrading from old version 3`] = `Object {}`;

exports[`transactionCache create starts fresh if upgrading from old version 4`] = `Object {}`;
